### PR TITLE
ENH: speed up rebin_piecewise_constant by vectorising loop

### PR DESCRIPTION
**Pull request by andyfaff**
TL:DR; the loop in `rebin_piecewise_constant` adds significantly to processing time. Vectorising achieves good speedups.

For a given processing task I have to call `rebin_piecewise_constant` around 2000 times. Each of the rebin calls has ~220 `x1` and `x2` bins. This means the rebinning loop was operating 440 000 times.  Profiling showed this to be a major slowdown in my code.  I decided to vectorise `rebin_piecewise_constant`.  This results in a massive speedup. With the following example:

    m = 10000
    n = 5001
    # bin edges 
    x_old = np.linspace(0., 1., m+1)
    x_new = np.linspace(-0.1, 1.2, n+1)

    # some arbitrary distribution
    y_old = 1. + np.sin(x_old[:-1]*np.pi) / np.ediff1d(x_old)

    %timeit rebin.rebin(x_old, y_old, x_new, interp_kind='piecewise_constant')

the original code takes 42.5 ms per loop. The vectorised code takes 410 microseconds per loop. This is a factor of 100 faster.

### CODE REVIEW

**Comment by jhykes** [comment]
100x speedup sounds good to me! Thanks for sharing your improvements.
